addons/addon-base-raas/packages/base-raas-services/lib/data-egress/data-egress-service.js

```diff
@@ -0,0 +1,326 @@
+    if (!s3Policy.Statement) {
+      s3Policy.Statement = [];
+    }
+    return { s3BucketName, s3Policy };
+  }
+
+  async addEgressStoreBucketPolicy(requestContext, egressStore, memberAccountId) {
+    const { s3BucketName, s3Policy } = await this.getS3BucketAndPolicy();
+
+    const statementParamFunctions = [];
+    if (egressStore.envPermission.read) {
+      statementParamFunctions.push(getStatementParamsFn);
+    }
+    if (egressStore.envPermission.write) {
+      statementParamFunctions.push(putStatementParamsFn);
+    }
+    if (egressStore.envPermission.read || egressStore.envPermission.write) {
+      statementParamFunctions.push(listStatementParamsFn);
+    }
+    const revisedStatements = await this.getRevisedS3Statements(
+      s3Policy,
+      egressStore,
+      s3BucketName,
+      statementParamFunctions,
+      oldStatement => this.addAccountToStatement(oldStatement, memberAccountId),
+    );
+
+    await this.updateS3BucketPolicy(s3BucketName, s3Policy, revisedStatements);
+
+    // Write audit event
+    await this.audit(requestContext, { action: 'add-egress-store-bucket-policy', body: s3Policy });
+  }
+
+  addAccountToStatement(oldStatement, memberAccountId) {
+    const principal = this.getRootArnForAccount(memberAccountId);
+    const statement = this.addEmptyPrincipalIfNotPresent(oldStatement);
+    if (Array.isArray(statement.Principal.AWS)) {
+      // add the principal if it doesn't exist already
+      if (!statement.Principal.AWS.includes(principal)) {
+        statement.Principal.AWS.push(principal);
+      }
+    } else if (statement.Principal.AWS !== principal) {
+      statement.Principal.AWS = [statement.Principal.AWS, principal];
+    }
+    return statement;
+  }
+
+  // @private
+  getRootArnForAccount(memberAccountId) {
+    return `arn:aws:iam::${memberAccountId}:root`;
+  }
+
+  // @private
+  addEmptyPrincipalIfNotPresent(statement) {
+    if (!statement.Principal) {
+      statement.Principal = {};
+    }
+    if (!statement.Principal.AWS) {
+      statement.Principal.AWS = [];
+    }
+    return statement;
+  }
+
+  // @private
+  async getRevisedS3Statements(s3Policy, egressStore, bucket, statementParamFunctions, updateStatementFn) {
+    const revisedStatementsPerStudy = _.map(statementParamFunctions, statementParameterFn => {
+      const statementParams = statementParameterFn(bucket, egressStore.prefix);
+      let oldStatement = s3Policy.Statement.find(statement => statement.Sid === statementParams.statementId);
+      if (!oldStatement) {
+        oldStatement = this.createAllowStatement(
+          statementParams.statementId,
+          statementParams.actions,
+          statementParams.resource,
+          statementParams.condition,
+        );
+      }
+      const newStatement = updateStatementFn(oldStatement);
+      return newStatement;
+    });
+    return revisedStatementsPerStudy;
+  }
+
+  // @private
+  async updateS3BucketPolicy(s3BucketName, s3Policy, revisedStatements) {
```

Comment:
Is it possible to move some of these methods to a utility class. They seem to be copied from environment-resource-service

Reply:
Moved some reusable functions into utils. Commit ID: a71878eb3a7938ebb93504e4179ff52e3d77baab